# Trabajo práctico - Programación Lineal


Una refinería produce tres productos, combustible para aviones, combustible para vehículos y kerosene. Gracias a los buenos precios de venta que mantiene, la refinería vende todos los productos que produce. El proceso de producción se compone de refinado, fraccionado y embalaje. Para realizar estas tareas, la empresa cuenta con 4 sectores:

* refinado: capacidad mensual de 38.000 horas, gasto fijo de $5.000.000

* fraccionado: capacidad mensual de 80.000 horas, gasto fijo de $5.000.000

* embalaje de combustible para aviones: capacidad mensual de 4.000 horas, gasto fijo de $2.000.000

* embalaje de combustible para vehículos: capacidad mensual de 6.000 horas, gasto fijo de $1.000.000

* embalaje de kerosene: capacidad mensual de 7.000 horas, gasto fijo de $500.000



El tiempo requerido para refinar 1000 litros de combustible para aviones es de 10 horas, mientras que para fraccionarlos son necesarias 20 horas y 4 para su embalaje. Para el combustible para vehículos los tiempos son de 5, 10 y 2 horas respectivamente, y para los 1000 litros de kerosene de 3, 6 y 1 hora.


El precio de venta es de \$16000 para los mil litros de combustible para aviones. \$8000 para los mil litros de combustible para vehículos. \$4000 los mil litros de kerosene. El costo de la materia prima para mil litros de combustible para aviones es de \$4000, el de refinado de \$4100, el de fraccionado de \$1000, y el de embalaje de \$1000. Para 1000 litros de combustible para vehículos, los costos son de \$1000, \$3000, \$600 y \$500, respectivamente. Y para 1000 litros de kerosene de \$500, \$1500, \$400 y \$400.


En el último mes, la refinería produjo 500.000 litros de combustible para aviones, 3.000.000 litros de combustible para vehículos y 6.000.000 litros de kerosene, lo que implicó una pérdida en el combustible para aviones al considerar prorrateados los gastos fijos.

El gerente de ventas señaló que estudios de mercado indican que no es posible aumentar el precio de venta del combustible para aviones y, por lo tanto, para aumentar las ganancias de la empresa se debe discontinuar su producción.
Sin embargo, la conclusión del jefe del departamento de embalaje de combustible para aviones es distinta. Según él, los gastos fijos de su sector inciden tanto en el costo de cada litro de combustible para aviones porque se produce
poco de este producto, y para aumentar las ganancias de la empresa propone que se produzca más.

En vista de las distintas propuestas, el director de la empresa necesita ayuda para evaluarlas y tomar las decisiones apropiadas. Para esto, analizar los siguientes ítems, evitando reoptimizar en todos los casos que sea posible:

In [ ]:
!pip install cplex

##1.

Calcular la ganancia o pérdida (prorrateando los gastos fijos) de cada producto que se obtuvo en el mes
anterior (cuando se produjeron 500.000 litros de combustible para aviones, 3.000.000 de combustible para
vehículos y 6.000.000 litros de kerosene) y la ganancia (o pérdida) total de la compañía.

In [ ]:
gastosfijorefinado = 5000000
gastofijofraccionado = 5000000
gastofijoembalajeAviones = 2000000
gastofijoembalajeVehiculos = 1000000
gastofijoembalajeKeroseno = 500000

precioventaAviones = 16000
precioventaVehiculos = 8000
precioventaKeroseno = 4000

costoCombustibleAv = 4000
costoRefinadoAv = 4100
costoFraccionadoAv = 1000
costoEmbalajeAv = 1000

costoCombustibleVe = 1000
costoRefinadoVe = 3000
costoFraccionadoVe = 600
costoEmbalajeVe = 500

costoCombustibleKer = 500
costoRefinadoKer = 1500
costoFraccionadoKer = 400
costoEmbalajeKer = 400

costoAviones = costoCombustibleAv + costoRefinadoAv + costoFraccionadoAv + costoEmbalajeAv
costoVehiculos = costoCombustibleVe + costoRefinadoVe + costoFraccionadoVe + costoEmbalajeVe
costoKeroseno = costoCombustibleKer + costoRefinadoKer + costoFraccionadoKer + costoEmbalajeKer

produccionAviones = 500000
produccionVehiculos = 3000000
produccionKeroseno = 6000000

horasrefinadototal = 10*produccionAviones + 5*produccionVehiculos + 3*produccionKeroseno
horasfraccionadototal =  20*produccionAviones + 10*produccionVehiculos + 6*produccionKeroseno


prorrateogastofijoaviones = (10*produccionAviones/horasrefinadototal)* gastosfijorefinado + (20*produccionAviones/horasfraccionadototal)* gastofijofraccionado
prorrateogastofijoVehiculos = (5*produccionVehiculos/horasrefinadototal)* gastosfijorefinado + (10*produccionVehiculos/horasfraccionadototal)* gastofijofraccionado
prorrateogastofijoKeroseno = (3*produccionKeroseno/horasrefinadototal)* gastosfijorefinado + (6*produccionKeroseno/horasfraccionadototal)* gastofijofraccionado

gananciaAviones = (precioventaAviones - costoAviones)*produccionAviones/1000 - gastofijoembalajeAviones - prorrateogastofijoaviones
gananciaVehiculos = (precioventaVehiculos - costoVehiculos)*produccionVehiculos/1000 - gastofijoembalajeVehiculos - prorrateogastofijoVehiculos
gananciaKeroseno = (precioventaKeroseno - costoKeroseno)*produccionKeroseno/1000 - gastofijoembalajeKeroseno - prorrateogastofijoKeroseno
gananciaTotal = gananciaAviones + gananciaVehiculos + gananciaKeroseno

print("Ganancia Aviones: ",gananciaAviones)
print("Ganancia Vehiculos: ",gananciaVehiculos)
print("Ganancia Keroseno: ",gananciaKeroseno)
print("Ganancia Total: ",gananciaTotal)

Ganancia Aviones:  -365789.47368421056
Ganancia Vehiculos:  3752631.5789473685
Ganancia Keroseno:  1963157.8947368423
Ganancia Total:  5350000.0


##2.
Si la empresa no hubiese producido combustible para aviones manteniendo en los mismos valores los otros productos, ¿la ganancia de la compañía habría sido mejor? Suponer que se cierra el sector de embalaje de  combustibles para aviones.

In [ ]:
produccionAviones = 0
gastofijoembalajeAviones = 0
horasrefinadototal = 10*produccionAviones + 5*produccionVehiculos + 3*produccionKeroseno
horasfraccionadototal =  20*produccionAviones + 10*produccionVehiculos + 6*produccionKeroseno

prorrateogastofijoaviones = (10*produccionAviones/horasrefinadototal)* gastosfijorefinado + (20*produccionAviones/horasfraccionadototal)* gastofijofraccionado
prorrateogastofijoVehiculos = (5*produccionVehiculos/horasrefinadototal)* gastosfijorefinado + (10*produccionVehiculos/horasfraccionadototal)* gastofijofraccionado
prorrateogastofijoKeroseno = (3*produccionKeroseno/horasrefinadototal)* gastosfijorefinado + (6*produccionKeroseno/horasfraccionadototal)* gastofijofraccionado

gananciaAviones = (precioventaAviones - costoAviones)*produccionAviones/1000 - gastofijoembalajeAviones - prorrateogastofijoaviones
gananciaVehiculos = (precioventaVehiculos - costoVehiculos)*produccionVehiculos/1000 - gastofijoembalajeVehiculos - prorrateogastofijoVehiculos
gananciaKeroseno = (precioventaKeroseno - costoKeroseno)*produccionKeroseno/1000 - gastofijoembalajeKeroseno - prorrateogastofijoKeroseno
gananciaTotal = gananciaAviones + gananciaVehiculos + gananciaKeroseno

print("Ganancia Aviones: ",gananciaAviones)
print("Ganancia Vehiculos: ",gananciaVehiculos)
print("Ganancia Keroseno: ",gananciaKeroseno)
print("Ganancia Total: ",gananciaTotal)

Ganancia Aviones:  0.0
Ganancia Vehiculos:  3154545.454545455
Ganancia Keroseno:  1245454.5454545459
Ganancia Total:  4400000.000000001


Podemos observar que si bien el combustible de aviones ya "no producen perdida", los gastos fijos de refinado y fraccionado ahora recaen únicamente sobre el keroseno y el combustible para vehículos. Lo que impacta sobre la ganancia de estos artículos reduciendola, y finalmente la ganancia total es menor.

## 3.
¿Y si hubiese aumentado lo máximo posible la producción de los otros productos? Suponer que se cierra el
sector de embalaje de combustibles para aviones.

In [ ]:
import cplex

In [ ]:
# 1. Inicializar CPLEX
prob = cplex.Cplex()

# 2. Definir el tipo de problema
prob.objective.set_sense(prob.objective.sense.maximize)

# 3. Definir las variables
nombres_variables = ["ca", "cv", "ck"]
cotas_superiores = [0, cplex.infinity, cplex.infinity]
prob.variables.add(obj=[16000 - 10100, 8000-5100, 4000-2800], ub=cotas_superiores, names=nombres_variables) #

# 4. Definir las restricciones
nombres_restricciones = ["Purificado", "Fraccionado","embalajeAutos","embalajeVehiculos","embalajeKerosen"]
filas = [
    [["ca", "cv","ck"], [10, 5,3]],  # 10ca + 5cv + 3ck <= 38000
    [["ca", "cv","ck"], [20, 10,6]],    #  20ca +   10cv + 6ck <= 80000
    [["ca", "cv","ck"], [4, 0, 0]],     # 4ca 4000
    [["ca", "cv","ck"], [0, 2, 0]],
    [["ca", "cv","ck"], [0, 0, 1]],
]
sentidos = ["L", "L","L","L","L"]  # L: <=, G: >=, E: =
lados_derechos = [38000, 80000,4000,6000,7000]
prob.linear_constraints.add(lin_expr=filas, senses=sentidos, rhs=lados_derechos, names=nombres_restricciones)

# Para depurar el codigo
prob.write('ejemplo1.lp')

# 5. Resolver el problema
prob.solve()

# 6. Mostrar la solucion
print('--------------------------------------------------')
print("Estado de la solución:", prob.solution.get_status_string(status_code = prob.solution.get_status()))
print("Valor objetivo:", prob.solution.get_objective_value())
print("Solución:", prob.solution.get_values())

Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
LP Presolve eliminated 5 rows and 3 columns.
All rows and columns eliminated.
Presolve time = 0.00 sec. (0.00 ticks)
--------------------------------------------------
Estado de la solución: optimal
Valor objetivo: 17100000.0
Solución: [0.0, 3000.0, 7000.0]


In [ ]:
gananciatotal = 17100000-11500000
print(gananciatotal)

5600000


después de descontar los costos fijos de refinado, fraccionado y embalaje tanto de combustible para vehiculos como de kerosen, la ganancia final que nos queda es $5600000. Efectivamente esta ganancia es mayor a la que teníamos antes.


##4.
Determinar la cantidad óptima de producción mensual de cada producto para maximizar la ganancia de la
compañía.


In [ ]:
# 1. Inicializar CPLEX
prob = cplex.Cplex()

# 2. Definir el tipo de problema
prob.objective.set_sense(prob.objective.sense.maximize)

# 3. Definir las variables
nombres_variables = ["ca", "cv", "ck"]
cotas_superiores = [cplex.infinity, cplex.infinity, cplex.infinity]
prob.variables.add(obj=[16000 - 10100, 8000-5100, 4000-2800], ub=cotas_superiores, names=nombres_variables) #restando a la ganancia los costos de materia prima

# 4. Definir las restricciones
nombres_restricciones = ["Refinado", "Fraccionado","embalajeAutos","embalajeVehiculos","embalajeKerosen"]
filas = [
    [["ca", "cv","ck"], [10, 5,3]],  # 10ca + 5cv + 3ck <= 38000
    [["ca", "cv","ck"], [20, 10,6]],    #  20ca +   10cv + 6ck <= 80000
    [["ca", "cv","ck"], [4, 0, 0]],     # 4ca <= 4000
    [["ca", "cv","ck"], [0, 2, 0]],     #2cv <= 6000
    [["ca", "cv","ck"], [0, 0, 1]],    #ck <= 7000
]
sentidos = ["L", "L","L","L","L"]  # L: <=, G: >=, E: =
lados_derechos = [38000, 80000,4000,6000,7000]
prob.linear_constraints.add(lin_expr=filas, senses=sentidos, rhs=lados_derechos, names=nombres_restricciones)

# Para depurar el codigo
prob.write('ejemplo1.lp')

# 5. Resolver el problema
prob.solve()

# 6. Mostrar la solucion
print('--------------------------------------------------')
print("Estado de la solución:", prob.solution.get_status_string(status_code = prob.solution.get_status()))
print("Valor objetivo:", prob.solution.get_objective_value())
print("Solución:", prob.solution.get_values())

Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
LP Presolve eliminated 4 rows and 0 columns.
Reduced LP has 1 rows, 3 columns, and 3 nonzeros.
Presolve time = 0.01 sec. (0.00 ticks)

Iteration log . . .
Iteration:     1   Dual objective     =      19800000.000000
--------------------------------------------------
Estado de la solución: optimal
Valor objetivo: 19800000.0
Solución: [1000.0, 3000.0, 4333.333333333333]


In [ ]:
gananciatotal = 19800000 - 13500000
print(gananciatotal)

6300000


La ganancia total, luego de descontar los gastos, es de $6.300.000, que es la mayor hasta el momento

##5.
Indicar al director estas cantidades, el costo por 1000 litros de cada producto
(prorrateando los costos fijos) y la ganancia total de la empresa.

In [ ]:
# Producción óptima (resultado del punto 4)
produccionAviones = 1000
produccionVehiculos = 3000
produccionKeroseno = 4333.333333333

# Costos variables por 1000 L
costoAviones = 4000
costoVehiculos = 3000
costoKeroseno = 1000

# Precios de venta por 1000 L
precioAviones = 12000
precioVehiculos = 10000
precioKeroseno = 5000

# Gasto fijo común
gastoFijoRefinado = 5000000
gastoFijoFraccionado = 5000000

# Gasto fijo específico de embalaje
gastoFijoEmbalajeAviones = 500000
gastoFijoEmbalajeVehiculos = 1000000
gastoFijoEmbalajeKeroseno = 1500000

# Horas necesarias por 1000 L
refAviones = 10
refVehiculos = 5
refKeroseno = 3
fracAviones = 20
fracVehiculos = 10
fracKeroseno = 6

# Cálculo de horas totales utilizadas por cada producto
horasRefAviones = produccionAviones  / refAviones
horasRefVehiculos = produccionVehiculos / refVehiculos
horasRefKeroseno = produccionKeroseno / refKeroseno
totalHorasRefinado = horasRefAviones + horasRefVehiculos + horasRefKeroseno

horasFracAviones = produccionAviones / fracAviones
horasFracVehiculos = produccionVehiculos / fracVehiculos
horasFracKeroseno = produccionKeroseno / fracKeroseno
totalHorasFraccionado = horasFracAviones + horasFracVehiculos + horasFracKeroseno

# Prorrateo de gastos fijos comunes
prorrateoRefAviones = gastoFijoRefinado * horasRefAviones / totalHorasRefinado
prorrateoRefVehiculos = gastoFijoRefinado * horasRefVehiculos / totalHorasRefinado
prorrateoRefKeroseno = gastoFijoRefinado * horasRefKeroseno / totalHorasRefinado

prorrateoFracAviones = gastoFijoFraccionado * horasFracAviones / totalHorasFraccionado
prorrateoFracVehiculos = gastoFijoFraccionado * horasFracVehiculos / totalHorasFraccionado
prorrateoFracKeroseno = gastoFijoFraccionado * horasFracKeroseno / totalHorasFraccionado

# Costos totales por 1000 litros
costototalAviones_1000L = (
    costoAviones +
    (prorrateoRefAviones + prorrateoFracAviones + gastoFijoEmbalajeAviones) / produccionAviones
)

costototalVehiculos_1000L = (
    costoVehiculos +
    (prorrateoRefVehiculos + prorrateoFracVehiculos + gastoFijoEmbalajeVehiculos) / produccionVehiculos
)

costototalKeroseno_1000L = (
    costoKeroseno +
    (prorrateoRefKeroseno + prorrateoFracKeroseno + gastoFijoEmbalajeKeroseno) / produccionKeroseno
)

# Ganancias por producto
gananciaAviones = (precioAviones - costototalAviones_1000L) * produccionAviones / 1000
gananciaVehiculos = (precioVehiculos - costototalVehiculos_1000L) * produccionVehiculos / 1000
gananciaKeroseno = (precioKeroseno - costototalKeroseno_1000L) * produccionKeroseno / 1000
gananciaTotal = gananciaAviones + gananciaVehiculos + gananciaKeroseno

print("\n--- Costos totales por cada 1000 litros (con prorrateo) ---")
print("Costo Aviones por 1000 L: ", round(costototalAviones_1000L, 2))
print("Costo Vehículos por 1000 L: ", round(costototalVehiculos_1000L, 2))
print("Costo Keroseno por 1000 L: ", round(costototalKeroseno_1000L, 2))

print("\n--- Producción óptima ---")
print("Aviones: ", produccionAviones)
print("Vehículos: ", produccionVehiculos)
print("Keroseno: ", round(produccionKeroseno, 2))

print("\n--- Ganancias ---")
print("Ganancia Aviones: ", round(gananciaAviones, 2))
print("Ganancia Vehículos: ", round(gananciaVehiculos, 2))
print("Ganancia Keroseno: ", round(gananciaKeroseno, 2))
print("Ganancia Total: ", round(gananciaTotal, 2))



--- Costos totales por cada 1000 litros (con prorrateo) ---
Costo Aviones por 1000 L:  4966.32
Costo Vehículos por 1000 L:  4265.98
Costo Keroseno por 1000 L:  2900.56

--- Producción óptima ---
Aviones:  1000
Vehículos:  3000
Keroseno:  4333.33

--- Ganancias ---
Ganancia Aviones:  7033.68
Ganancia Vehículos:  17202.07
Ganancia Keroseno:  9097.58
Ganancia Total:  33333.33


##6.
Al escuchar esto, el gerente de producción propuso aumentar la producción contratando 500 horas extras al mes del personal del sector de fraccionado. Asesorar al director sobre esta propuesta.

Sabemos que el diccionario optimo tiene la siguiente forma matricial:

$Xb^* = B^{-1*}b - B^{-1*} An^*Xn$

$z^* = Cb^*B^{-1*}b + (Cn^* - Cb^*B^{-1*}An^*)Xn$


In [ ]:
# Access the solution object and then the basis object to get the basis status
base = prob.solution.basis.get_basis()
primal_values = prob.solution.get_values()
dual_values = prob.solution.get_dual_values()
costos_reducidos = prob.solution.get_reduced_costs()

# You can then print or use the basis_status which is a tuple containing
# the basis status for variables and linear constraints
print("Base (Variables, Restricciones Slack):", base)
print("Valores Primales:", primal_values)
print("Valores Duales:", dual_values)
print("Costos Reducidos:", costos_reducidos)

Base (Variables, Restricciones Slack): ([1, 1, 1], [0, 1, 0, 0, 1])
Valores Primales: [1000.0, 3000.0, 4333.333333333333]
Valores Duales: [400.0, -0.0, 475.0000000000001, 450.0000000000001, -0.0]
Costos Reducidos: [0.0, 0.0, 0.0]


Al aumentar las horas del sector de fraccionado, estamos aumentando nuestra segunda restricción. Lo que estamos modificando en nuestra forma matricial es el vector  $b$   $( b_2 )$. Como estamos aumentando las restricciones, nuestra región factible está aumentando, es decir todas las soluciones que antes eran factibles lo siguen siendo, y ahora hay más soluciones factibles. Esto nos dice que nuestra solución óptima también sigue siendo factible. Como no cambiamos la función objetivo, entonces deducimos que al aumentar una restricción nuestro valor objetivo solo puede o aumentar o mantenerse igual.

También podemos observar que el valor dual asociado a la segunda restricción $y_2 = 0$. Esto nos dice que si aumentamos o disminuimos poco esa restricción, no va a afectar a nuestro valor objetivo, pues esta restricción actualmente no está siendo saturada.

Por todo esto podríamos aconsejarle al gerente que contratar horas extras en el sector de fraccionado no nos va a dar más ganancias actualmente.

De todas maneras vamos a terminar de analizar la información que tenemos disponible y armar el diccionario optimo para todo el resto del análisis.

Podemos ver que las 3 variables (ca, cv, ck) estan en la base. Ademas podemos ver que la variables slack asociadas a la segunda y quinta restriccion tambien estan en la base ($x_5,x_8$).

Es decir que los valores primales = $Xb^* = [1000, 3000, 4333.333, x_5, x_8]$

Tambien observamos que:

Costos reducidos = $(Cn^* - Cb^*B^{-1*}An^*) = [0,0,0]$

Valores duales/solución dual = $Cb^*B^{-1*} = [400, 0, 475, 450, 0]$

$B^* =
 \begin{pmatrix}
  10 & 5 & 3 & 0 & 0 \\
  20 & 10 & 6 & 1 & 0 \\
  4  & 0  & 0 & 0 & 0  \\
  0 & 2 & 0 & 0 & 0 \\
  0 & 0 & 1 & 0 & 1
 \end{pmatrix}
 A_n^* =
 \begin{pmatrix}
  1 & 0 & 0  \\
  0 & 0 & 0 \\
  0  & 1  & 0   \\
  0 & 0 & 1  \\
  0 & 0 & 0
 \end{pmatrix}
b =\begin{pmatrix}
 38000\\
 80000\\
 4000\\
 6000\\
 7000
 \end{pmatrix}
C_b = \begin{pmatrix}
 5.900\\
 2.900\\
 1.200\\
 0\\
 0
 \end{pmatrix}
 C_n = \begin{pmatrix}
 0\\
 0\\
 0
 \end{pmatrix}$

.

Teniendo $B$ calculamos $B^{-1}$

$B^{-1*} =
 \begin{pmatrix}
  0 & 0 & 1/4 & 0 & 0 \\
  0 & 0 & 0 & 1/2 & 0 \\
  1/3  & 0  & -5/6 & -5/6 & 0  \\
  -2 & 1 & 0 & 0 & 0 \\
  -1/3 & 0 & 5/6 & 5/6 & 1
 \end{pmatrix} $

Y tambien deducimos el valor de $x_5,x_8$:

$Xb^* = [1000, 3000, 4333.333, 4000, 2666.666]$

Al aumentar las horas del sector de fraccionado, estamos aumentando nuestra segunda restriccion. Lo que estamos modificando en nuestra forma matricial es el vector $b$ ($b_2$), y esto podria generar que nuestro diccionario no sea mas factible, si $X_b = B^{-1*}b < 0$.

$B^{-1}\begin{pmatrix}
 38000\\
 b2\\
 4000\\
 6000\\
 7000
 \end{pmatrix} =
  \begin{pmatrix}
  0 & 0 & 1/4 & 0 & 0 \\
  0 & 0 & 0 & 1/2 & 0 \\
  1/3  & 0  & -5/6 & -5/6 & 0  \\
  -2 & 1 & 0 & 0 & 0 \\
  -1/3 & 0 & 5/6 & 5/6 & 1
 \end{pmatrix}
 \begin{pmatrix}
 38000\\
 b2\\
 4000\\
 6000\\
 7000
 \end{pmatrix}
 $

 Podemos ver que la unica opcion de que $b2$ haga infactible a nuestro diccionario es si:

 $b_2 -2*38000 < 0 $ --> $b_2 < 76000$

In [ ]:
# 1. Inicializar CPLEX
prob = cplex.Cplex()

# 2. Definir el tipo de problema
prob.objective.set_sense(prob.objective.sense.maximize)

# 3. Definir las variables
nombres_variables = ["ca", "cv", "ck"]
cotas_superiores = [cplex.infinity, cplex.infinity, cplex.infinity]
prob.variables.add(obj=[16000 - 10100, 8000-5100, 4000-2800], ub=cotas_superiores, names=nombres_variables) #restando a la ganancia los costos de materia prima

# 4. Definir las restricciones
nombres_restricciones = ["Refinado", "Fraccionado","embalajeAutos","embalajeVehiculos","embalajeKerosen"]
filas = [
    [["ca", "cv","ck"], [10, 5,3]],  # 10ca + 5cv + 3ck <= 38000
    [["ca", "cv","ck"], [20, 10,6]],    #  20ca +   10cv + 6ck <= 80000
    [["ca", "cv","ck"], [4, 0, 0]],     # 4ca <= 4000
    [["ca", "cv","ck"], [0, 2, 0]],     #2cv <= 6000
    [["ca", "cv","ck"], [0, 0, 1]],    #ck <= 7000
]
sentidos = ["L", "L","L","L","L"]  # L: <=, G: >=, E: =
lados_derechos = [38000, 80500,4000,6000,7000] # Aqui aumentamos 80000 a 80500 (horas de fraccionado)
prob.linear_constraints.add(lin_expr=filas, senses=sentidos, rhs=lados_derechos, names=nombres_restricciones)

# Para depurar el codigo
prob.write('ejemplo1.lp')

# 5. Resolver el problema
prob.solve()

# 6. Mostrar la solucion
print('--------------------------------------------------')
print("Estado de la solución:", prob.solution.get_status_string(status_code = prob.solution.get_status()))
print("Valor objetivo:", prob.solution.get_objective_value())
print("Solución:", prob.solution.get_values())

Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
LP Presolve eliminated 4 rows and 0 columns.
Reduced LP has 1 rows, 3 columns, and 3 nonzeros.
Presolve time = 0.00 sec. (0.00 ticks)

Iteration log . . .
Iteration:     1   Dual objective     =      19800000.000000
--------------------------------------------------
Estado de la solución: optimal
Valor objetivo: 19800000.0
Solución: [1000.0, 3000.0, 4333.333333333333]


##7.
Otra propuesta del gerente es contratar 1000 horas extras al mes del personal del sector de refinado. Indicar
al director si es conveniente aceptar esta nueva propuesta y hasta cuánto debería pagar por cada hora extra
de este sector.


Como dijimos en el punto anterior al  aumentar una restricción,  la función objetivo puede o aumentar o mantenerse igual. En este caso queremos modificar la primer restricción y a diferencia del punto anterior, en este caso podemos observar que la variable $y_1$ asociada a esta restricción vale 400. Es decir que por cada unidad que aumentemos la primer restricción, nuestra función objetivo aumentara en 400, así que en principio si nos sirve contratar horas extras.

In [ ]:
# 1. Inicializar CPLEX
prob = cplex.Cplex()

# 2. Definir el tipo de problema
prob.objective.set_sense(prob.objective.sense.maximize)

# 3. Definir las variables
nombres_variables = ["ca", "cv", "ck"]
cotas_superiores = [cplex.infinity, cplex.infinity, cplex.infinity]
prob.variables.add(obj=[16000 - 10100, 8000-5100, 4000-2800], ub=cotas_superiores, names=nombres_variables) #restando a la ganancia los costos de materia prima

# 4. Definir las restricciones
nombres_restricciones = ["Refinado", "Fraccionado","embalajeAutos","embalajeVehiculos","embalajeKerosen"]
filas = [
    [["ca", "cv","ck"], [10, 5,3]],  # 10ca + 5cv + 3ck <= 38000
    [["ca", "cv","ck"], [20, 10,6]],    #  20ca +   10cv + 6ck <= 80000
    [["ca", "cv","ck"], [4, 0, 0]],     # 4ca <= 4000
    [["ca", "cv","ck"], [0, 2, 0]],     #2cv <= 6000
    [["ca", "cv","ck"], [0, 0, 1]],    #ck <= 7000
]
sentidos = ["L", "L","L","L","L"]  # L: <=, G: >=, E: =
lados_derechos = [39000, 80500,4000,6000,7000] # Aqui aumentamos 80000 a 80500 (horas de fraccionado) y de 38000 a 39000 (horas de refinado)
prob.linear_constraints.add(lin_expr=filas, senses=sentidos, rhs=lados_derechos, names=nombres_restricciones)

# Para depurar el codigo
prob.write('ejemplo1.lp')

# 5. Resolver el problema
prob.solve()

# 6. Mostrar la solucion
print('--------------------------------------------------')
print("Estado de la solución:", prob.solution.get_status_string(status_code = prob.solution.get_status()))
print("Valor objetivo:", prob.solution.get_objective_value())
print("Solución:", prob.solution.get_values())

Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
LP Presolve eliminated 4 rows and 0 columns.
Reduced LP has 1 rows, 3 columns, and 3 nonzeros.
Presolve time = 0.00 sec. (0.00 ticks)

Iteration log . . .
Iteration:     1   Dual objective     =      20200000.000000
--------------------------------------------------
Estado de la solución: optimal
Valor objetivo: 20200000.0
Solución: [1000.0, 3000.0, 4666.666666666667]


In [ ]:
ganancia_final = 20200000 - 19800000
valor_por_hora_extra = ganancia_final / 1000
print("Valor de cada hora extra: $",ganancia_final)
print("La empresa aumenta su ganancia total en: $",valor_por_hora_extra)

Valor de cada hora extra: $ 400000
La empresa aumenta su ganancia total en: $ 400.0


La propuesta de contratar 1000 horas extras en el sector de refinado es conveniente, ya que permite aumentar las ganancias en 400,000 mensuales. Por lo tanto, la empresa podría aceptar esta propuesta si el costo por hora extra es menor o igual a 400. Cualquier monto mayor haría que la inversión no sea rentable.

##8.
Si el director decide pagar por hora extra la mitad del valor máximo indicado en el punto anterior, ¿en cuánto
aumentaría la ganacia mensual de la compañía?


In [ ]:
# Datos conocidos
ganancia_extra_bruta = 400000   # Incremento bruto en la ganancia por horas extras (del punto anterior)
horas_extra = 1000              # Horas extras contratadas
valor_hora_max = 400            # Valor máximo por hora extra (punto anterior)

pago_hora_extra = valor_hora_max / 2

costo_horas_extra = horas_extra * pago_hora_extra

ganancia_neta_extra = ganancia_extra_bruta - costo_horas_extra

print(f"Si se paga {pago_hora_extra} por hora extra, la ganancia mensual aumentará en: ${ganancia_neta_extra:.2f}")


Si se paga 200.0 por hora extra, la ganancia mensual aumentará en: $200000.00


##9.
Por otro lado, el gerente de compras propone cambiar algunos proveedores, lo que permitiría bajar el costo
de la materia prima del aceite para vehículos de \$1000 a \$800 por cada 1000 litros procesados. ¿Cambiaría el
plan de producción óptimo? Si es así, dar la nueva planificación óptima.


Queremos saber si al bajar el costo de materia prima de **cv** de $\$1000$ a $\$800$ cambia el plan óptimo encontrado anteriormente.

Conocemos la ganancia unitaria de cada producto
- ca: \$16000 – (4000 + 4100 + 1000 + 1000) = **\$4900**
- cv: \$8000 – (1000 + 3000 + 600 + 500) = **\$2900** (pasa a **\$3100**)
- ck: \$4000 – (500 + 1500 + 400 + 400) = **\$1200**

Sabemos que las restricciones activas son las de embalaje:
- Embalaje $c_{a}$: 4 horas
- Embalaje $c_{v}$: 2 horas
- Embalaje $c_{k}$: 1 hora

Entonces:
$B = \begin{bmatrix}
4 & 0 & 0 \\
0 & 2 & 0 \\
0 & 0 & 1
\end{bmatrix}
\quad \Rightarrow \quad
B^{-1} = \begin{bmatrix}
1/4 & 0 & 0 \\
0 & 1/2 & 0 \\
0 & 0 & 1
\end{bmatrix}$

Columnas de $A_{N}$ (refinado y fraccionado)

$A_{N} = \begin{bmatrix}
10 & 20 \\
5 & 10 \\
3 & 6
\end{bmatrix}$

Ganancias básicas actualizadas :

$C_{B}$ = [4900, **3100**, 1200]

Cálculo de coeficientes reducidos :

$\bar{c}_{j} = c_{j} - C_{B} \cdot B^{-1} \cdot A_{N}$

Sabemos que si todos los coeficientes reducidos se mantienen negativos, el óptimo no cambia.

|                         | Refinado | Fraccionado |
|-------------------------|----------|-------------|
| $\bar{c}_{j}$ antes   | –25600   | –51200      |
| $\bar{c}_{j}$ después | –26100   | –52200      |

Los coeficientes reducidos siguen siendo negativos, por lo tanto la solución óptima se mantiene.El valor del optimo si se modifica pero no la solucion optima.    
No es necesario reoptimizar.

##10.
Y si se modificara el proceso de refinado de kerosene para bajar de \$1500 a \$900 por cada 1000 litros procesado,
¿cambiaría el plan óptimo? Si es así, dar la nueva planificación óptima.


busco  $C_3$ tal que  $(C_N - CbB^{-1}A_N)$ <= 0 donde $C_b = (5900,2900,C_3,0,0)$ para que la solucion siga factible.

  $(0,0,0) - (5900,2900,C_3,0,0)*B^{-1}A_N <= (0,0,0)$

$B^{-1*}A_n =
 \begin{pmatrix}
  0 & 0 & 1/4 & 0 & 0 \\
  0 & 0 & 0 & 1/2 & 0 \\
  1/3  & 0  & -5/6 & -5/6 & 0  \\
  -2 & 1 & 0 & 0 & 0 \\
  -1/3 & 0 & 5/6 & 5/6 & 1
 \end{pmatrix}
 \begin{pmatrix}
  1 & 0 & 0  \\
  0 & 0 & 0 \\
  0  & 1  & 0   \\
  0 & 0 & 1  \\
  0 & 0 & 0
 \end{pmatrix}
 =
  \begin{pmatrix}
  0 & 1/4 & 0  \\
  0 & 0 & 1/2 \\
  1/3  & -5/6  & -5/6   \\
  -2 & 0 & 0  \\
  -1/3 & 5/6 & 5/6
 \end{pmatrix}
 $


  
 $C_b B^{-1*}A_n = (C_3/3, 1475 - 5*C_3/6, 1450 - 5*C_3/6) $

 Entonces $(-C_3/3, -1475 + 5*C_3/6, -1450 + 5*C_3/6) < (0,0,0)$
 $ \implies  (C_3 > 0 , C_3 < 1770 , C_3 < 1740) $

 $\implies 0< C_3 < 1740$

Pero el nuevo costo de Keroseno es (4000 - 2200) = 1800 > 1740 . Por lo tanto si se modificara la solucion optima, veamoslo con el siguiente codigo :

In [ ]:
import cplex

# Datos originales
costoAviones = 16000 - 10100
costoVehiculos = 8000 - 5100

# Nuevo costo refinado kerosene (2200 en vez de 2800 (1500-900 = 600 de diferencia en el total))
costoKeroseno = 4000 - 2200

# Crear modelo
prob = cplex.Cplex()
prob.objective.set_sense(prob.objective.sense.maximize)

# Variables y coeficientes de la función objetivo actualizados
nombres_variables = ["ca", "cv", "ck"]
obj_coef = [costoAviones, costoVehiculos, costoKeroseno]
prob.variables.add(obj=obj_coef, names=nombres_variables, ub=[cplex.infinity]*3)

filas = [
    [["ca", "cv","ck"], [10, 5, 3]],  # Purificado
    [["ca", "cv","ck"], [20, 10, 6]], # Fraccionado
    [["ca", "cv","ck"], [4, 0, 0]],   # Embalaje Aviones
    [["ca", "cv","ck"], [0, 2, 0]],   # Embalaje Vehículos
    [["ca", "cv","ck"], [0, 0, 1]]    # Embalaje Keroseno
]
sentidos = ["L", "L", "L", "L", "L"]
lados_derechos = [38000, 80000, 4000, 6000, 7000]

prob.linear_constraints.add(lin_expr=filas, senses=sentidos, rhs=lados_derechos)

# Resolver
prob.solve()

# Mostrar resultados
print('--------------------------------------------------')
print("Estado de la solución:", prob.solution.get_status_string())
print("Valor objetivo:", prob.solution.get_objective_value())
print("Nueva planificación óptima (litros):", prob.solution.get_values())


Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
LP Presolve eliminated 4 rows and 0 columns.
Reduced LP has 1 rows, 3 columns, and 3 nonzeros.
Presolve time = 0.00 sec. (0.00 ticks)

Iteration log . . .
Iteration:     1   Dual objective     =      22560000.000000
--------------------------------------------------
Estado de la solución: optimal
Valor objetivo: 22560000.0
Nueva planificación óptima (litros): [1000.0, 1399.999999999999, 7000.0]


##11.
La empresa está evaluando comenzar a procesar gasoil. El tiempo requerido para refinar 1000 litros de gasoil
es de 4 horas, mientras que para fraccionarlos son necesarias 8 horas y para su embalaje 1.5 horas. El costo
de la materia prima para mil litros de gasoil es de \$4000, el de refinado de \$4100, el de fraccionado de \$1000.
El embalaje de gasoil lo realizaría el sector de embalaje de kerosene. ¿Cuál debería ser el menor precio de
venta de los 1000 litros de gasoil para que su producción sea conveniente para la empresa?


Aqui lo que estamos haciendo es procesar un nuevo producto, es decir agregamos una variable a nuestro problema de programacion lineal, buscamos el menor precio de venta para que su producción sea conveniente. Es decir, si llamamos x4 a la nueva variable, queremos $C_4$ tal que  $(C_4 - CbB^{-1}A_4)$ > 0 para que sea rentrable su producción.
Recordemos que la nueva variable sera no basica, y $A_4 = (4,8,0,0,1.5)$

Entonces $C_4 > [400,0,475,450,0] * [4,8,0,0,1.5] = 1600$

$C_4$ representa la ganancia final del producto, asi que el precio final debe tener en cuenta los gastos, los cuales son por cada mil litros: \$4000 (materia prima) + \$4100 (refinado) + \$1000 (fraccionado) = \$9100

Por lo tanto, **el precio de venta final para el gasoil debe ser mayor a \$10700** para que sea rentable su producción.

##12.
 La empresa va a agregar un control de calidad a todos sus productos. Controlar los 1000 litros de combustible
para aviones requiere 5 horas, los de combustible para vehículos 3 horas y 2 horas los 1000 litros de kerosene.
Si el sector de control de calidad dispone de 20000 horas mensuales, ¿cambiaría el plan óptimo? Si es así, dar
la nueva planificación óptima.


El nuevo control de calidad impacta sobre nuestro modelo como una nueva restricción.

$(5Ca + 3CV + 2Ck) <= 20000 $

Y nuestra solución óptima es $Xb^* = [1000, 3000, 4333.333]$, la cual no cumple esta nueva restricción. Entonces tenemos que buscar la nueva planificación óptima.

In [ ]:
# 1. Inicializar CPLEX
prob = cplex.Cplex()

# 2. Definir el tipo de problema
prob.objective.set_sense(prob.objective.sense.maximize)

# 3. Definir las variables
nombres_variables = ["ca", "cv", "ck"]
cotas_superiores = [cplex.infinity, cplex.infinity, cplex.infinity]
prob.variables.add(obj=[16000 - 10100, 8000-5100, 4000-2800], ub=cotas_superiores, names=nombres_variables) #restando a la ganancia los costos de materia prima

# 4. Definir las restricciones
nombres_restricciones = ["Refinado", "Fraccionado","embalajeAutos","embalajeVehiculos","embalajeKerosen","controlCalidad"]
filas = [
    [["ca", "cv","ck"], [10, 5,3]],  # 10ca + 5cv + 3ck <= 38000
    [["ca", "cv","ck"], [20, 10,6]],    #  20ca +   10cv + 6ck <= 80000
    [["ca", "cv","ck"], [4, 0, 0]],     # 4ca <= 4000
    [["ca", "cv","ck"], [0, 2, 0]],     #2cv <= 6000
    [["ca", "cv","ck"], [0, 0, 1]],    #ck <= 7000
    [["ca", "cv","ck"], [5, 3, 2]]
]
sentidos = ["L", "L","L","L","L","L"]  # L: <=, G: >=, E: =
lados_derechos = [38000, 80000,4000,6000,7000,20000]
prob.linear_constraints.add(lin_expr=filas, senses=sentidos, rhs=lados_derechos, names=nombres_restricciones)

# Para depurar el codigo
prob.write('ejemplo1.lp')

# 5. Resolver el problema
prob.solve()

# 6. Mostrar la solucion
print('--------------------------------------------------')
print("Estado de la solución:", prob.solution.get_status_string(status_code = prob.solution.get_status()))
print("Valor objetivo:", prob.solution.get_objective_value())
print("Solución:", prob.solution.get_values())

Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
LP Presolve eliminated 4 rows and 0 columns.
Reduced LP has 2 rows, 3 columns, and 6 nonzeros.
Presolve time = 0.00 sec. (0.00 ticks)

Iteration log . . .
Iteration:     1   Dual objective     =      19800000.000000
--------------------------------------------------
Estado de la solución: optimal
Valor objetivo: 18200000.0
Solución: [1000.0, 3000.0, 3000.0]


In [ ]:
soloptima = prob.solution.get_values()
print("nueva solucion optima: ", soloptima)
gananciatotal = 18200000 - 13500000
print("ganancia total: $", gananciatotal)
perdida = 19800000 - 13500000 - gananciatotal
print("perdida (respecto a no tener control de calidad): $", perdida)


nueva solucion optima:  [1000.0, 3000.0, 0.0]
ganancia total: $ 4700000
perdida (respecto a no tener control de calidad): $ 1600000
